contracts/AelinPool.sol

```diff
@@ -385,6 +406,65 @@ contract AelinPool is AelinERC20, MinimalProxyFactory, IAelinPool {
         emit PurchasePoolToken(msg.sender, purchaseTokenAmount);
     }
 
+    /**
+     * @dev allows anyone to become a purchaser with a qualified erc721
+     * nft in the pool depending on the scenarios
+     *
+     * Scenarios:
+     * 1. each wallet holding a qualified NFT to deposit an unlimited amount of purchase tokens
+     * 2. certain amount of purchase tokens per wallet regardless of the number of qualified NFTs held
+     * 3. certain amount of Investment tokens per qualified NFT held
+     */
+    function purchasePoolScenario1and2for721(address _collectionAddress, uint256 _purchaseTokenAmount) public {
+        require(hasNftList, "pool does not have an NFT list");
+        require(block.timestamp < purchaseExpiry, "not in purchase window");
+        require(IERC721(_collectionAddress).balanceOf(msg.sender) > 0, "should own atleast 1 nft");
+
+        NftData memory tmpNftData = nftCollectionDetails[_collectionAddress];
+        require(tmpNftData.collectionAddress == _collectionAddress, "collection not in the pool");
+
+        if (tmpNftData.purchaseAmount > 0) {
+            require(
+                nftAllowList[msg.sender] + _purchaseTokenAmount <= tmpNftData.purchaseAmount,
+                "purchase amount should be less than the allocation"
+            );
+            nftAllowList[msg.sender] += _purchaseTokenAmount;
+        }
+
+        // is blacklisting required here? else, anyone can transfer nfts and purchase more
+
+        IERC20(purchaseToken).safeTransferFrom(msg.sender, address(this), _purchaseTokenAmount);
+        _mint(msg.sender, _purchaseTokenAmount);
+    }
+
+    function purchasePoolScenario3for721(
+        address _collectionAddress,
+        uint256[] calldata _tokenIds,
+        uint256 _purchaseAmountPerToken
+    ) public {
+        require(hasNftList, "pool does not have an NFT list");
+        require(block.timestamp < purchaseExpiry, "not in purchase window");
+        require(_tokenIds.length > 0, "number of token ids should be more than 0");
+
+        NftData memory tmpNftData = nftCollectionDetails[_collectionAddress];
+        require(tmpNftData.collectionAddress == _collectionAddress, "collection not in the pool");
+        require(tmpNftData.purchaseAmountPerToken, "purchaseAmount is not per token");
```

Comment:
this type of error would not be needed if we put all the purchase 721 into a single file, it will check the rules of the collection(s) you pass into the function and for each one it would auto know which rules based on what the sponsor set

Reply:
purchase721 is a tricky method since you can have scenarios 1, 2 and 3 across 3 different collections and you can have a purchaser holding all 3 types and you don't know which one they will call, if not multiple. 

E.g. maybe they have 10 BAYC NFTs and those have an allocation of 10 so 100 in total and then they have 15 MAYC and those have an allocation of 10 no matter how many you have in your wallet, they could call purchase721 for both collections and send in 110 in total. 

The scenario where you get unlimited per NFT should be the only one you can't call with any other scenarios passed in because there is no point in using more than 1 if you have unlimited allowance. 

Of course we could simplify this stuff by getting rid of one of the scenarios or only allowing 1 scenario per pool maybe. We still have some flexibility if it will make the code a lot simpler

Reply:
we can do multiple collectionAddresses at the same time but it will make the code a little messy. have a look at 96e358261f8e04f73bb908935ba0db1146f14740 and maybe we can discuss if that's fine